src/IB/IBFESurfaceMethod.cpp

```diff
@@ -1388,162 +2289,1758 @@ void IBFESurfaceMethod::endDataRedistribution(Pointer<PatchHierarchy<NDIM> > /*h
+            {
+
+               P_o_qp[local_indices[k]] =  P_j_qp[local_indices[k]] + Q_i_qp[local_indices[k]]; //0.5 * ( P_j_qp[local_indices[k]] + Q_i_qp[local_indices[k]] + Q_o_qp[local_indices[k]]);
+	           P_i_qp[local_indices[k]] =  -(P_j_qp[local_indices[k]] - Q_o_qp[local_indices[k]]);
+            }
+        }
+
+        // Loop over the elements and accumulate the right-hand-side values.
+        qrule.reset();
+        qp_offset = 0;
+        for (unsigned int e_idx = 0; e_idx < num_active_patch_elems; ++e_idx)
+        {
+            Elem* const elem = patch_elems[e_idx];
+
+            P_o_dof_map_cache.dof_indices(elem, P_o_dof_indices);
+            P_o_rhs_e.resize(static_cast<int>(P_o_dof_indices.size()));
+            
+            P_i_dof_map_cache.dof_indices(elem, P_i_dof_indices);
+            P_i_rhs_e.resize(static_cast<int>(P_i_dof_indices.size()));
+
+            for (unsigned int d = 0; d < NDIM; ++d)
+            {
+                X_dof_map.dof_indices(elem, X_dof_indices[d], d);
+            }
+            get_values_for_interpolation(x_node, *X_ghost_vec, X_dof_indices);
+            const bool qrule_changed =
+                FEDataManager::updateInterpQuadratureRule(qrule, d_default_interp_spec, elem, x_node, patch_dx_min);
+            if (qrule_changed) fe->attach_quadrature_rule(qrule.get());
+
+            fe->reinit(elem);
+
+            const unsigned int n_qp = qrule->n_points();
+            const size_t n_basis = X_dof_indices[0].size();
+            for (unsigned int qp = 0; qp < n_qp; ++qp)
+            {
+                const int idx = qp_offset + qp;
+                for (unsigned int k = 0; k < n_basis; ++k)
+                {
+                    const double p_JxW = phi[k][qp] * JxW[qp];
+                    P_o_rhs_e(k) += P_o_qp[idx] * p_JxW;
+                    P_i_rhs_e(k) += P_i_qp[idx] * p_JxW;
+                }
+            }
+
+            P_o_dof_map.constrain_element_vector(P_o_rhs_e, P_o_dof_indices);
+            P_o_rhs_vec->add_vector(P_o_rhs_e, P_o_dof_indices);
+            
+            P_i_dof_map.constrain_element_vector(P_i_rhs_e, P_i_dof_indices);
+            P_i_rhs_vec->add_vector(P_i_rhs_e, P_i_dof_indices);
+            
+            qp_offset += n_qp;
+        }
+    }
+    P_o_rhs_vec->close();
+	P_i_rhs_vec->close();
+	
+    d_fe_data_managers[part]->computeL2Projection(
+        *P_o_vec, *P_o_rhs_vec, P_O_SYSTEM_NAME, d_default_interp_spec.use_consistent_mass_matrix);
+
+    d_fe_data_managers[part]->computeL2Projection(
+        *P_i_vec, *P_i_rhs_vec, P_I_SYSTEM_NAME, d_default_interp_spec.use_consistent_mass_matrix);
+
+    d_X_half_vecs[part]->close();
+    d_X_current_vecs[part]->close();
+    d_X_new_vecs[part]->close();
+
+    d_P_o_half_vecs[part]->close();
+    d_P_i_half_vecs[part]->close();
+    d_P_j_IB_ghost_vecs[part]->close();
+    d_X_IB_ghost_vecs[part]->close();
+
+    return;
+
+} // interpolatePressureForTraction
```

Comment:
This function has the same logic everywhere for interior and exterior tractions. Could we convert some of this code to loops over interior and exterior data, instead of writing nearly the same thing for each traction?

Reply:
(well, almost everywhere!)

Reply:
That's true. I thought something can be done here or whenever we compute quantities on both sides of the IB. Mostly it seems all the computations are the same except for the sign of the normal vector that switches between the two....I'll work on fixing this. Thanks for the suggestion David.